src/main/java/net/gegy1000/plasmid/block/CustomBlock.java

```diff
@@ -0,0 +1,92 @@
+package net.gegy1000.plasmid.block;
+
+import com.google.common.base.Preconditions;
+import net.gegy1000.plasmid.registry.TinyRegistry;
+import net.minecraft.text.Text;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.BlockPos;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class CustomBlock {
+    private static final TinyRegistry<CustomBlock> REGISTRY = TinyRegistry.newStable();
+    private static final Map<BlockPos, CustomBlock> BLOCK_MAP = new HashMap<>();
+
+    private final Identifier id;
+    private final Text name;
+
+    private CustomBlock(Identifier identifier, Text name) {
+        this.id = identifier;
+        this.name = name;
+    }
+
+    public static Set<Identifier> getKeys() {
+        return REGISTRY.keySet();
+    }
+
+    @Nullable
+    public static CustomBlock get(Identifier identifier) {
+        return REGISTRY.get(identifier);
+    }
+
+    public boolean applyTo(BlockPos pos) {
+        if (this.name != null && !BLOCK_MAP.containsKey(pos)) {
+            BLOCK_MAP.put(pos, this);
+            return true;
+        }
+        return false;
+    }
+
+    public boolean setBlock(BlockPos pos) {
+        return this.applyTo(pos);
+    }
+
+    @Nullable
+    public static CustomBlock match(BlockPos pos) {
+        return BLOCK_MAP.get(pos);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
Is the equals method needed here? It could also use some whitespace cleanup around the if and else statements.

Reply:
got it.